### PR TITLE
feat(SPV-679) finalize contacts

### PR DESCRIPTION
**Pull request by chris-4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet-web-backend/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet-web-backend/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: config/viper.go [hunk 1]
```diff
@@ -82,3 +84,8 @@ func setWebsocketDefaults() {
 	viper.SetDefault(EnvWebsocketHistoryMax, 300)
 	viper.SetDefault(EnvWebsocketHistoryTtl, 10)
 }
+
+func setContactsDefaults() {
+	viper.SetDefault(EnvContactsPasscodePeriod, uint(3600)) //1h
+	viper.SetDefault(EnvContactsPasscodeDigits, uint(4))
```
Comment (arkadiuszos4chain) [inline config/viper.go:90]: I'm almost certain that if we demonstrate this to someone from BA, there will be a question: *Is it possible to use two digits for TOTP?*

File: transports/http/auth/auth_middleware.go [hunk 2]
```diff
@@ -49,24 +49,26 @@ func (h *AuthMiddleware) ApplyToApi(c *gin.Context) {
...
 		userId == nil ||
 		paymail == nil {
-		return nil, nil, nil, nil, ErrorUnauthorized
+		return nil, nil, nil, nil, nil, ErrorUnauthorized
 	}
 
 	err = h.checkAccessKey(accessKey.(string), accessKeyId.(string))
 	if err != nil {
-		return nil, nil, nil, nil, fmt.Errorf("%w: %s", ErrorUnauthorized, err) // we could use this with go 1.20: fmt.Errorf("%w: %w", ErrorUnauthorized, err)
+		return nil, nil, nil, nil, nil, fmt.Errorf("%w: %s", ErrorUnauthorized, err) // we could use this with go 1.20: fmt.Errorf("%w: %w", ErrorUnauthorized, err)
```
Comment (arkadiuszos4chain) [inline transports/http/auth/auth_middleware.go:71]: Can we use `fmt.Errorf("%w %w")`?

File: transports/http/endpoints/api/contacts/endpoints.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	if err != nil {
+		h.log.Error().Msgf("Invalid payload: %s", err)
+		c.JSON(http.StatusBadRequest, "Invalid request. Please check contact details")
+		return
+	}
+
+	passcode, err := h.cService.GenerateTotpForContact(c.Request.Context(), c.GetString(auth.SessionXPriv), &contact)
+	if err != nil {
+		h.log.Error().Msgf("An error occurred while generating TOTP for the contact: %s", err)
+		c.JSON(http.StatusBadRequest, "An error occurred while generating TOTP for the contact.")
```
Comment (arkadiuszos4chain) [inline transports/http/endpoints/api/contacts/endpoints.go:194]: IMO should be `InternalServerError`

File: transports/http/endpoints/api/contacts/endpoints.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	var req UpsertContact
+	err := c.Bind(&req)
+	if err != nil {
+		h.log.Error().Msgf("Invalid payload: %s", err)
+		c.JSON(http.StatusBadRequest, "Invalid request. Please check contact details")
+		return
+	}
+
+	_, err = h.cService.UpsertContact(c.Request.Context(), c.GetString(auth.SessionAccessKey), paymail, req.FullName, &req.Metadata)
+	if err != nil {
```
Comment (arkadiuszos4chain) [inline transports/http/endpoints/api/contacts/endpoints.go:91]: Differentiating errors from the server is probably overkill for the reference implementation, but if we don't do it, we should return `InternalServerError`, in my opinion

File: transports/http/endpoints/api/contacts/endpoints.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+//	@Summary Accept a contact
+//	@Tags contact
+//	@Produce json
+//	@Success 200
+//	@Router /api/v1/contact/accepted/{paymail} [patch]
+func (h *handler) acceptContact(c *gin.Context) {
+	paymail := c.Param("paymail")
+
+	err := h.cService.AcceptContact(c.Request.Context(), c.GetString(auth.SessionAccessKey), paymail)
+	if err != nil {
```
Comment (arkadiuszos4chain) [inline transports/http/endpoints/api/contacts/endpoints.go:111]: Differentiating errors from the server is probably overkill for the reference implementation, but if we don't do it, we should return `InternalServerError`, in my opinion

File: transports/http/auth/auth_middleware.go (hunk 2 above)
Comment (Nazarii-4chain) [inline transports/http/auth/auth_middleware.go:71]: Maybe we can remove this comment

**Review by Nazarii-4chain — DISMISSED** [review]

File: transports/http/auth/session_middleware.go [hunk 6]
```diff
@@ -19,11 +19,13 @@ const (
 	SessionAccessKey   = "accessKey"
 	SessionUserId      = "userId"
 	SessionUserPaymail = "paymail"
+	SessionXPriv       = "xPriv"
 )
 
 // NewSessionMiddleware create Session middleware that is retrieving auth token from cookie.
 func NewSessionMiddleware(db *sql.DB, engine *gin.Engine) router.ApiMiddlewareFunc {
-	store, err := postgres.NewStore(db, []byte("secret"))
+	secret := viper.GetString(config.EnvHttpServerSessionSecret)
```
Comment (arkadiuszos4chain) [inline transports/http/auth/session_middleware.go:27]: <3

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

File: config/viper.go (hunk 1 above)
Comment (chris-4chain) [reply config/viper.go:90]: So, what would you recommend?

File: transports/http/auth/auth_middleware.go (hunk 2 above)
Comment (chris-4chain) [reply transports/http/auth/auth_middleware.go:71]: Yes we can and we should 👍

File: transports/http/endpoints/api/contacts/endpoints.go (hunk 3 above)
Comment (chris-4chain) [reply transports/http/endpoints/api/contacts/endpoints.go:194]: You're totally right 👍

File: config/viper.go (hunk 1 above)
Comment (arkadiuszos4chain) [reply config/viper.go]: Let's use 2 digits - same pattern as in the mockups

Comment (chris-4chain) [reply config/viper.go]: Changed 👍 
Also adjusted in frontend

**Review by arkadiuszos4chain — APPROVED** [review]

**Review by Nazarii-4chain — APPROVED** [review]

**Review by dorzepowski — APPROVED** [review]

**Comment by wregulski** [comment]
testing

**Comment by wregulski** [comment]
One thing to discuss later -> if we want to allow user to add him/herself as a contact. Currently it's possible.

**Pull request merged**